In [1]:
import pandas as pd
import numpy as np

df = pd.read_csv('CatalogoProductos.csv', encoding='utf-8')
df

,ID_PRODUCTO,NOMBRE,CATEGORIA,MARCA
0,B0BW2LV57K,"Bocina Sonos Era 300 - Black - Wireless, Alexa...",Electronics,Sonos
1,B0B76W7JD9,Memoria RAM SystemBuild Evolution Camberly Fra...,Kitchen,SystemBuild
2,B0F53C5ZXW,Pantalla TCL 98 Inch Class QM8K Series Smart TV,Electronics,TCL
3,B0049B1VKY,Libro electrónico The Sixties: Diaries 1960°,Kindle Edition,Amazon Publishing
4,B0DF69NTJG,Libro electrónico Thirty Pieces of Silver (Spi...,Kindle Edition,Amazon Publishing
...,...,...,...,...
157,B09V3CPSCQ,Accesorio automotriz Westin 57-23955 HDX 3-Pie...,Electronics,Westin
158,B01N0F5LNK,Libro electrónico The Beeswax Workshop: How to...,Electronics,Amazon Publishing
159,B08DMLZQXW,Producto alimenticio Twinings Holiday Variety ...,Personal Computers,Twinings
160,B0DYQFW6TR,Televisión LG 77-Inch Class OLED evo AI 4K C5 ...,Kitchen,LG


In [2]:
print(df.describe())

       ID_PRODUCTO                                             NOMBRE  \
count          162                                                162   
unique         162                                                162   
top     B0BW2LV57K  Bocina Sonos Era 300 - Black - Wireless, Alexa...   
freq             1                                                  1   

          CATEGORIA              MARCA  
count           162                162  
unique           25                 85  
top     Electronics  Amazon Publishing  
freq             35                 31  


In [3]:
lista_categoria = np.unique(df.CATEGORIA)
print(lista_categoria)
print(len(lista_categoria))

['Apparel' 'Audio CD' 'Automotive' 'Blu -ray' 'Electronics' 'Escolar'
 'Grocery' 'Health and Beauty' 'Home' 'Kindle Edition' 'Kitchen'
 'Lawn & Patio' 'Misc.' 'Office Product' 'Personal Computers'
 'Product Bundle' 'Software Download' 'Sports' 'Target'
 'Tools & Home Improvement' 'Toy' 'Vi nyl' 'Video Game' 'Watch'
 'Wireless Phone Accessory']
25


Vemos que categorias como 'Blu -ray' y 'Vi nyl' estan separadas, entonces hacemos un remplace para corregir las categorias

In [4]:
df['CATEGORIA'] = df['CATEGORIA'].replace({'Blu -ray': 'Blu-ray', 'Vi nyl': 'Vinyl'})
np.unique(df.CATEGORIA)

array(['Apparel', 'Audio CD', 'Automotive', 'Blu-ray', 'Electronics',
       'Escolar', 'Grocery', 'Health and Beauty', 'Home',
       'Kindle Edition', 'Kitchen', 'Lawn & Patio', 'Misc.',
       'Office Product', 'Personal Computers', 'Product Bundle',
       'Software Download', 'Sports', 'Target',
       'Tools & Home Improvement', 'Toy', 'Video Game', 'Vinyl', 'Watch',
       'Wireless Phone Accessory'], dtype=object)

Ya que han quedado las categorias vamos a ver si hay algun problema con las marcas

In [5]:
lista_marca = np.unique(df.MARCA)
print(lista_marca)
print(len(lista_marca))

['AMD' 'ATN' 'Alesis' 'Amazon' 'Amazon Publishing' 'Apple' 'BISSELL'
 'Bali & Pari' 'Bestop' 'Blue Buffalo' 'Bose' 'Breville' 'Carlisle'
 'Chefman' 'Core Power' 'Corsair' 'Cosori' 'Coway' 'Crucial' 'Cuisinart'
 'CyberPower' 'DEWALT' 'DJI' 'Darigold' 'Dell' 'Denon' 'Design Toscano'
 'EGO POWER+' 'EGO Power+' 'FDW' 'Fitbit' 'Furinno' 'GIGABYTE' 'Garmin'
 'Garvee' 'GarveeLife' 'Generic' 'Genesis' 'Hallmark' 'IRONCK'
 "Internet's Best" 'Kimberly-Clark' 'Kingston' 'LEGO' 'LG' 'Lenovo'
 'Leviton' 'Lion Brand' 'Marc Jacobs' 'Martin Svensson Home'
 'Microsoft Xbox' 'NOCO' 'Nilight' 'Ninja' 'Nintendo' 'Optimum Nutrition'
 'Panasonic' 'Pokemon' 'PowerColor' 'Pro Comp' 'Quicken' 'SONGMICS'
 'Samsung' 'Seagate' 'Sonos' 'Sony' 'Sun Zero' 'Switch Sticks'
 'SystemBuild' 'TCL' 'Temu' 'The Stupell Home Decor Collection'
 'Trademark Innovations' 'Twinings' 'Under Armour' 'Universal Music'
 'Urban Accents' 'WD_BLACK' 'WORX' 'Warner Bros.' 'Western Digital'
 'Westin' 'Whynter' 'YouTheFan' 'Zojirushi']
85


La mayoria de las marcas estan bien, pero estan por ejemplo 'EGO POWER+' 'EGO Power+' que son la misma, o  'Garvee' 'GarveeLife' o  'Amazon' 'Amazon Publishing'. Entonces esas son las que vamos a normalizar

In [6]:
df['MARCA'] = df['MARCA'].replace({'EGO Power+': 'EGO POWER+', 'Garvee': 'GarveeLife', 'Amazon Publishing':'Amazon' })
np.unique(df.MARCA)

array(['AMD', 'ATN', 'Alesis', 'Amazon', 'Apple', 'BISSELL',
       'Bali & Pari', 'Bestop', 'Blue Buffalo', 'Bose', 'Breville',
       'Carlisle', 'Chefman', 'Core Power', 'Corsair', 'Cosori', 'Coway',
       'Crucial', 'Cuisinart', 'CyberPower', 'DEWALT', 'DJI', 'Darigold',
       'Dell', 'Denon', 'Design Toscano', 'EGO POWER+', 'FDW', 'Fitbit',
       'Furinno', 'GIGABYTE', 'Garmin', 'GarveeLife', 'Generic',
       'Genesis', 'Hallmark', 'IRONCK', "Internet's Best",
       'Kimberly-Clark', 'Kingston', 'LEGO', 'LG', 'Lenovo', 'Leviton',
       'Lion Brand', 'Marc Jacobs', 'Martin Svensson Home',
       'Microsoft Xbox', 'NOCO', 'Nilight', 'Ninja', 'Nintendo',
       'Optimum Nutrition', 'Panasonic', 'Pokemon', 'PowerColor',
       'Pro Comp', 'Quicken', 'SONGMICS', 'Samsung', 'Seagate', 'Sonos',
       'Sony', 'Sun Zero', 'Switch Sticks', 'SystemBuild', 'TCL', 'Temu',
       'The Stupell Home Decor Collection', 'Trademark Innovations',
       'Twinings', 'Under Armour', 'Universal M

Tambien vamos a estandarizar los nombres para ello vamos a pasarlos todos a minusculas y a quitar acentos, esto lo vamos a hacer en los nombres y categorias, ya que los nombre de las empresas estan registrados entonces se deben de escribir tal cual

In [7]:
reemplazos = {
    r'[áàäâ]': 'a',
    r'[éèëê]': 'e',
    r'[íìïî]': 'i',
    r'[óòöô]': 'o',
    r'[úùüû]': 'u',
    r'ñ': 'n'
}

columnas = ['CATEGORIA', 'NOMBRE']

for col in columnas:
    # 1. Convertir a minúsculas primero
    df[col] = df[col].str.lower()
    
    # 2. Reemplazar acentos usando regex
    for patron, reemplazo in reemplazos.items():
        df[col] = df[col].replace(patron, reemplazo, regex=True)

Ahora que ya tenemos el catalogo corregido lo exportamos

In [8]:
df.to_csv('CatalogoProductosEstandar.csv', index=False)


# Insercion de valores #

Ahora lo que vamos a hacer es agregarles un precio actual, un stock y un precio de fabricacion

In [10]:
#Iniciamos primero por el precio, para ello vamos a traer el valor mas reciente del csv de HistorialPrecios
df = pd.read_csv('HistorialPrecios.csv', encoding='utf-8')
#Tambien traemos el nuevo catalogo que creamos
df_catalogo = pd.read_csv('CatalogoProductosEstandar.csv', encoding='utf-8')

df


,ID_PRODUCTO,FECHA,PRECIO_REGISTRADO
0,B0BZHTVHN5,2025-01-03 00:00:00,158.03
1,B0BZHTVHN5,2025-01-04 00:00:00,158.03
2,B0BZHTVHN5,2025-01-05 00:00:00,158.03
3,B0BZHTVHN5,2025-01-06 00:00:00,158.33
4,B0BZHTVHN5,2025-01-07 00:00:00,158.33
...,...,...,...
195897,B07N3PQM5N,2021-09-12 00:00:00,524.76
195898,B07N3PQM5N,2021-09-13 00:00:00,522.99
195899,B07N3PQM5N,2021-09-14 00:00:00,522.99
195900,B07N3PQM5N,2021-09-15 00:00:00,522.99


In [11]:
#Aseguramos que la fecha realmente sea tipo fecha
df['FECHA'] = pd.to_datetime(df['FECHA'])
# Ordenar por fecha descendente (la más reciente primero)
df_ordenado = df.sort_values(by='FECHA', ascending=False)

# Eliminar duplicados de ID_PRODUCTO, manteniendo solo la primera ocurrencia (la más reciente)
precio_reciente = df_ordenado.drop_duplicates(subset='ID_PRODUCTO', keep='first')

#Ahora lo que hacemos un merge mediante el ID del producto
# Suponiendo que tu df principal de productos se llama 'df_productos'
df_final = df_catalogo.merge(precio_reciente[['ID_PRODUCTO', 'PRECIO_REGISTRADO']], 
                              on='ID_PRODUCTO', 
                              how='left')
#Renombramos la columna para que se llame precio actual
df_final = df_final.rename(columns={'PRECIO_REGISTRADO': 'PRECIO_ACTUAL'})

df_final

,ID_PRODUCTO,NOMBRE,CATEGORIA,MARCA,PRECIO_ACTUAL
0,B0BW2LV57K,"bocina sonos era 300 - black - wireless, alexa...",electronics,Sonos,402.32
1,B0B76W7JD9,memoria ram systembuild evolution camberly fra...,kitchen,SystemBuild,107.14
2,B0F53C5ZXW,pantalla tcl 98 inch class qm8k series smart tv,electronics,TCL,3671.71
3,B0049B1VKY,libro electronico the sixties: diaries 1960°,kindle edition,Amazon,13.07
4,B0DF69NTJG,libro electronico thirty pieces of silver (spi...,kindle edition,Amazon,0.98
...,...,...,...,...,...
157,B09V3CPSCQ,accesorio automotriz westin 57-23955 hdx 3-pie...,electronics,Westin,790.97
158,B01N0F5LNK,libro electronico the beeswax workshop: how to...,electronics,Amazon,5.63
159,B08DMLZQXW,producto alimenticio twinings holiday variety ...,personal computers,Twinings,23.48
160,B0DYQFW6TR,television lg 77-inch class oled evo ai 4k c5 ...,kitchen,LG,2087.57


Ahora para el stock vamos a asignar productos que esten entre 0 y 100

In [12]:
df_final['STOCK'] = np.random.randint(0, 101, size=len(df_final))
df_final

,ID_PRODUCTO,NOMBRE,CATEGORIA,MARCA,PRECIO_ACTUAL,STOCK
0,B0BW2LV57K,"bocina sonos era 300 - black - wireless, alexa...",electronics,Sonos,402.32,48
1,B0B76W7JD9,memoria ram systembuild evolution camberly fra...,kitchen,SystemBuild,107.14,15
2,B0F53C5ZXW,pantalla tcl 98 inch class qm8k series smart tv,electronics,TCL,3671.71,52
3,B0049B1VKY,libro electronico the sixties: diaries 1960°,kindle edition,Amazon,13.07,77
4,B0DF69NTJG,libro electronico thirty pieces of silver (spi...,kindle edition,Amazon,0.98,10
...,...,...,...,...,...,...
157,B09V3CPSCQ,accesorio automotriz westin 57-23955 hdx 3-pie...,electronics,Westin,790.97,45
158,B01N0F5LNK,libro electronico the beeswax workshop: how to...,electronics,Amazon,5.63,30
159,B08DMLZQXW,producto alimenticio twinings holiday variety ...,personal computers,Twinings,23.48,73
160,B0DYQFW6TR,television lg 77-inch class oled evo ai 4k c5 ...,kitchen,LG,2087.57,31


In [13]:
#Ya por ultimo asignamos un precio de fabricacion, este precio debe ser menor al precio de venta, asi que vamos a asignar un porcentaje entre 10 y 20 por ciento, esto para que esta sea la reduccion

#Generamos un valor que va a ser el porcentaje, este va a estar entre el 80 y 90 %
porcentaje_reduccion = np.random.uniform(0.80, 0.90, size=len(df_final))

#Creamos la nueva columna, multiplicando el precio actual por su reduccion en porcentaje
df_final['PRECIO_FABRICACION'] = df_final['PRECIO_ACTUAL'] * porcentaje_reduccion

# Redondear
df_final['PRECIO_FABRICACION'] = df_final['PRECIO_FABRICACION'].round(2)
df_final

,ID_PRODUCTO,NOMBRE,CATEGORIA,MARCA,PRECIO_ACTUAL,STOCK,PRECIO_FABRICACION
0,B0BW2LV57K,"bocina sonos era 300 - black - wireless, alexa...",electronics,Sonos,402.32,48,356.75
1,B0B76W7JD9,memoria ram systembuild evolution camberly fra...,kitchen,SystemBuild,107.14,15,96.29
2,B0F53C5ZXW,pantalla tcl 98 inch class qm8k series smart tv,electronics,TCL,3671.71,52,3072.79
3,B0049B1VKY,libro electronico the sixties: diaries 1960°,kindle edition,Amazon,13.07,77,11.24
4,B0DF69NTJG,libro electronico thirty pieces of silver (spi...,kindle edition,Amazon,0.98,10,0.79
...,...,...,...,...,...,...,...
157,B09V3CPSCQ,accesorio automotriz westin 57-23955 hdx 3-pie...,electronics,Westin,790.97,45,646.44
158,B01N0F5LNK,libro electronico the beeswax workshop: how to...,electronics,Amazon,5.63,30,4.65
159,B08DMLZQXW,producto alimenticio twinings holiday variety ...,personal computers,Twinings,23.48,73,20.92
160,B0DYQFW6TR,television lg 77-inch class oled evo ai 4k c5 ...,kitchen,LG,2087.57,31,1827.97


Ya por ultimo lo exportamos

In [14]:
df_final.to_csv('Productos.csv', index=False,encoding='utf-8' )


# Analisis de Outliers #

Ahora que ya tenemos los productos y todas sus categorias normalizadas, nos falta checar algo importante, y es ver si no hay outliers en el precio. 
Hay que saber que estos precios no son estaticos, si no dinamicos, entonces van a ir cambiando, y tampoco necesariamente deben mantenerse en un incremento constante, ya que los precios son reales, entonces realmente un producto puede aumentar de precio por la inflacion, por guerras o por sucesos mundiales.
Por esta razón, no podemos asumir que cualquier variación grande sea un error, siendo por esto que nos vamos a centrar mas en  aquellos registros cuyo comportamiento se aleja significativamente de la tendencia histórica del propio producto, ya que estos valores atípicos podrían corresponder a errores de captura o anomalidades que sean mas de la plataforma y no tanto del producto.

In [24]:
df_historial = pd.read_csv("HistorialPrecios.csv")
#Nos aseguramos que realmente sea de formato fecha
df_historial["FECHA"] = pd.to_datetime(df_historial["FECHA"])

#Ahora medimos realmente el cambio porcentual por dia
df_historial = df_historial.sort_values(["ID_PRODUCTO", "FECHA"])
#Calculamos la variacion porcentual respecto al dia anterior
df_historial["CAMBIO_PCT"] = (df_historial.groupby("ID_PRODUCTO") ["PRECIO_REGISTRADO"] .pct_change())

df_historial

,ID_PRODUCTO,FECHA,PRECIO_REGISTRADO,CAMBIO_PCT
144695,B00007J5U7,2020-01-01,207.24,NaN
144696,B00007J5U7,2020-01-02,207.24,0.000000
144697,B00007J5U7,2020-01-03,207.24,0.000000
144698,B00007J5U7,2020-01-04,207.24,0.000000
144699,B00007J5U7,2020-01-05,207.24,0.000000
...,...,...,...,...
166987,B0GT7CRZ93,2026-05-23,229.98,0.000000
35896,B0GYY1F1R3,2026-05-20,80.98,NaN
166988,B0GYY1F1R3,2026-05-21,80.99,0.000123
166989,B0GYY1F1R3,2026-05-22,80.98,-0.000123


Como vemos la variacion del precio realmente no es tanta entre un dia y otro, entonces podemos filtrar como valores correctos a los valores que su variacion es cero, y revisar solamente los que realmente tengan una variacion considerable

In [28]:
#Calculamos que pocertanje tienen una variacion de cero
porcentaje_ceros = ((df_historial["CAMBIO_PCT"] == 0).mean()* 100)

print(f"Porcentaje de cambios iguales a 0: "f"{porcentaje_ceros:.2f}%"
)

Porcentaje de cambios iguales a 0: 92.96%


Es decir que solo el 8% realmente tiene un cambio porcentual, aunque no todos son realmente significativos, ya que pueden haber algunos en los que tengan un cambio del 1%, y como ya es diferente de 0 por eso se registra

In [30]:
df_historial["CAMBIO_PCT"].describe(
    percentiles=[
        0.01,
        0.05,
        0.10,
        0.25,
        0.50,
        0.75,
        0.90,
        0.95,
        0.99
    ]
)

count    195741.000000
mean          0.001935
std           0.069879
min          -0.724900
1%           -0.116893
5%            0.000000
10%           0.000000
25%           0.000000
50%           0.000000
75%           0.000000
90%           0.000000
95%           0.000000
99%           0.131410
max           2.812500
Name: CAMBIO_PCT, dtype: float64

Los resultados muestran una distribución altamente concentrada en cambios nulos. Aproximadamente el 95% de los registros mantienen exactamente el mismo precio respecto al registro anterior, mientras que únicamente el 1% supera una variación del 13.14%.

In [39]:
#Entonces lo que hacemos es sacar una media movil de los utlimos 7 dias

df_historial["MEDIA_MOVIL"] = (
    df_historial
    .groupby("ID_PRODUCTO")
    ["PRECIO_REGISTRADO"]
    .transform(
        lambda x: x.rolling(
            window=7,
            center=True,
            min_periods=3
        ).mean()
    )
)

In [41]:
#Vemos que tan lejos esta el valor de la media movil
df_historial["DESVIO_PCT"] = (
    abs(
        df_historial["PRECIO_REGISTRADO"]
        -
        df_historial["MEDIA_MOVIL"]
    )
    /
    df_historial["MEDIA_MOVIL"]
)

In [42]:
df_historial["DESVIO_PCT"].describe(
    percentiles=[
        0.90,
        0.95,
        0.99,
        0.995,
        0.999
    ]
)

count    195902.000000
mean          0.012646
std           0.043286
min           0.000000
50%           0.000000
90%           0.032276
95%           0.077816
99%           0.241318
99.5%         0.289846
99.9%         0.421590
max           1.172811
Name: DESVIO_PCT, dtype: float64

Eesto nos dice apenas el 99.5% de todo el total de los datos apenas presenta una variacion de 28% respecto a su media, entonces los registros que debemos de eliminar son aquellos que superen el 30%

In [43]:
#Establecemos la cota del 30% para que ese sea nuestro valor minimo
umbral = 0.30
outliers = df_historial[df_historial["DESVIO_PCT"] > umbral]

In [44]:
print(f"Outliers encontrados: {len(outliers)}")

print(
    f"Porcentaje eliminado: "
    f"{100 * len(outliers) / len(df_historial):.3f}%"
)

Outliers encontrados: 883
Porcentaje eliminado: 0.451%


In [46]:
#Borramos los outliers
df_limpio = df_historial.drop(outliers.index)
#Comprobamos resultados
print("Registros originales:", len(df_historial))
print("Outliers:", len(outliers))
print("Registros limpios:", len(df_limpio))

Registros originales: 195902
Outliers: 883
Registros limpios: 195019


In [47]:
#Guardamos csv
df_limpio.to_csv(
    "HistorialPrecios_EDA.csv",
    index=False
)
